src/PactNet.Native/MessageScenarioBuilder.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Collections.Generic;
+
+namespace PactNet.Native
+{
+    /// <summary>
+    /// Defines the message scenario builder
+    /// </summary>
+    public class MessageScenarioBuilder : IMessageScenarioBuilder
+    {
+        /// <summary>
+        /// Handles to create a new scenario with scenario builder
+        /// </summary>
+        public static IMessageScenarioBuilder NewScenario => new MessageScenarioBuilder();
+
+        /// <summary>
+        /// temporary description
+        /// </summary>
+        private string _descriptionAdded;
+
+        /// <summary>
+        /// tag to know if we are creating a scenario
+        /// </summary>
+        private bool _settingScenario;
+
+        private MessageScenarioBuilder()
+        {
+            Scenarios.All ??= new Dictionary<string, Func<dynamic>>();
+        }
+
+        /// <inheritdoc />
+        public IMessageScenarioBuilder WhenReceiving(string description)
+        {
+            try
+            {
+                ValidateDescription(description);
+
+                _settingScenario = true;
+                _descriptionAdded = description;
+
+                return this;
+            }
+            catch (Exception)
+            {
+                ClearScenario();
+                throw;
+            }
+        }
+
+        /// <inheritdoc />
+        public void WillPublishMessage(Func<dynamic> action)
```

Comment:
Is there any reason this can't have an overload which just takes a `dynamic` directly instead of requiring a `Func` every time?

Reply:
No, you need a function because you need to test the return of an algorithm and not just an object. In PACT messaging, the most important thing is not to test that an object is the same type as the one in the contract but that the object _pushed in the queue_ is the same. Therefore you would need a way to simulate the pushing of an object.